ydb/core/fq/libs/row_dispatcher/format_handler/filters/filters_set.cpp

```diff
@@ -8,236 +10,238 @@ namespace {
+        if (requestIter == InFlightCompilations_.end()) {
+            return TStatus::Fail(EStatusId::INTERNAL_ERROR, TStringBuilder() << "Compile response ignored for id " << cookie);
+        }
 
-        IPurecalcFilter::TPtr purecalcFilter;
-        if (const auto& predicate = filter->GetWhereFilter()) {
-            LOG_ROW_DISPATCHER_TRACE("Create purecalc filter for predicate '" << predicate << "' (filter id: " << filter->GetFilterId() << ")");
+        const auto [clientId, name] = requestIter->second;
 
-            auto filterStatus = CreatePurecalcFilter(filter);
-            if (filterStatus.IsFail()) {
-                return filterStatus;
-            }
-            purecalcFilter = filterStatus.DetachResult();
+        const auto handlersIter = CompileHandlers_.find(clientId);
+        if (handlersIter == CompileHandlers_.end()) {
+            return TStatus::Fail(EStatusId::INTERNAL_ERROR, TStringBuilder() << "Compile response ignored for id " << cookie << ", program with client id " << clientId << " not found");
         }
 
-        const auto [it, inserted] = Filters.insert({filter->GetFilterId(), TFilterHandler(*this, filter, std::move(purecalcFilter))});
-        if (!inserted) {
-            return TStatus::Fail(EStatusId::INTERNAL_ERROR, TStringBuilder() << "Failed to create new filter, filter with id " << filter->GetFilterId() << " already exists");
+        const auto iter = handlersIter->second.find(name);
+        if (iter == handlersIter->second.end()) {
+            return TStatus::Fail(EStatusId::INTERNAL_ERROR, TStringBuilder() << "Compile response ignored for id " << cookie << ", program with client id " << clientId << " not found");
         }
 
-        it->second.CompileFilter();
-        return TStatus::Success();
+        const auto result = iter->second;
+        handlersIter->second.erase(iter);
+        if (handlersIter->second.empty()) {
+            CompileHandlers_.erase(handlersIter);
+        }
+        InFlightCompilations_.erase(requestIter);
+        Counters_.InFlightCompileRequests->Dec();
+        return result;
     }
 
-    void RemoveFilter(NActors::TActorId filterId) override {
-        LOG_ROW_DISPATCHER_TRACE("Remove filter with id " << filterId);
+    TValueStatus<IProgramRunHandler::TPtr> AddRunProgram(TString name, IProcessedDataConsumer::TPtr consumer, IProgramHolder::TPtr programHolder) {
+        const auto clientId = consumer->GetClientId();
 
-        const auto it = Filters.find(filterId);
-        if (it == Filters.end()) {
-            return;
+        auto runHandler = CreateProgramRunHandler(name, std::move(consumer), std::move(programHolder));
+        auto& runHandlers = RunHandlers_[clientId];
+        const auto [iter, inserted] = runHandlers.emplace(name, std::move(runHandler));
+        if (!inserted) {
+            return TStatus::Fail(EStatusId::INTERNAL_ERROR, TStringBuilder() << "Failed to run new program, program with client id " << clientId << " already exists");
```

Comment:
Тут можно наверное ещё name добавить в ошибку


```suggestion
            return TStatus::Fail(EStatusId::INTERNAL_ERROR, TStringBuilder() << "Failed to run new program, program with client id " << clientId << " and name " << name << " already exists");
```